validator/src/journal/block_manager.rs

```diff
@@ -0,0 +1,448 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::collections::HashMap;
+use std::rc::Rc;
+use std::sync::RwLock;
+
+use block::Block;
+use journal::{BlockStore, NULL_BLOCK_IDENTIFIER};
+
+#[derive(Debug, PartialEq)]
+pub enum BlockManagerError {
+    MissingPredecessor(String),
+    MissingPredecessorInBranch,
+    MissingInput,
+}
+
+struct Anchor {
+    pub blockstore: String,
+    pub ref_count: i64,
+    pub block_id: String,
+}
+
+#[derive(Default)]
+pub struct BlockManager {
+    block_by_block_id: Rc<RwLock<HashMap<String, (Block, i64)>>>,
+    blockstore_by_name: Rc<RwLock<HashMap<String, Box<BlockStore>>>>,
+
+    anchors: Rc<RwLock<HashMap<String, Anchor>>>,
+}
+
+/// The BlockManager maintains integrity of all the blocks it contains,
+/// such that for any Block within the BlockManager,
+/// that Block's predecessor is also within the BlockManager.
+impl BlockManager {
+    pub fn new() -> Self {
+        BlockManager {
+            block_by_block_id: Rc::new(RwLock::new(HashMap::new())),
+            blockstore_by_name: Rc::new(RwLock::new(HashMap::new())),
+
+            anchors: Rc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    fn garbage_collect(&mut self) {
```

Comment:
I realize now that this garbage_collect function isn't written right. For every block that is removed, the previous_block_id block should have it's ref-count reduced.